File: src/paradigm.js
'use strict';

class ParadigmJudge {
  constructor(name, evaluateFn, weight = 1.0) {
    this.name = name;
    this._evaluate = evaluateFn;
    this.weight = weight;
  }

  evaluate(organism) {
    return this._evaluate(organism);
  }
}

class ParadigmPanel {
  constructor() {
    this._judges = [];
  }

  addJudge(judge) {
    this._judges.push(judge);
    return this;
  }

  removeJudge(name) {
    this._judges = this._judges.filter((j) => j.name !== name);
    return this;
  }

  getJudge(name) {
    return this._judges.find((j) => j.name === name) ?? null;
  }

  get judges() {
    return [...this._judges];
  }

  evaluate(organism) {
    if (this._judges.length === 0) return 0;
    let totalWeight = 0;
    let totalScore = 0;
    const scores = {};
    for (const judge of this._judges) {
      const score = judge.evaluate(organism);
      scores[judge.name] = score;
      totalScore += score * judge.weight;
      totalWeight += judge.weight;
    }
    const aggregate = totalWeight > 0 ? totalScore / totalWeight : 0;
    return { scores, aggregate };
  }

  evaluatePopulation(population) {
    for (const org of population) {
      const result = this.evaluate(org);
      org.fitness = result.aggregate;
      org.meta.paradigmScores = result.scores;
    }
    return population;
  }
}

module.exports = { ParadigmJudge, ParadigmPanel };

File: python/phyloid_engine/paradigm.py
from __future__ import annotations

from typing import Any, Callable

from .organism import Organism


class ParadigmJudge:
    """Single-objective fitness function with a name and weight."""

    def __init__(
        self, name: str, evaluate_fn: Callable[[Organism], float], weight: float = 1.0
    ) -> None:
        self.name = name
        self._evaluate = evaluate_fn
        self.weight = weight

    def evaluate(self, organism: Organism) -> float:
        return self._evaluate(organism)


class ParadigmPanel:
    """Multi-objective fitness aggregator using weighted judges."""

    def __init__(self) -> None:
        self._judges: list[ParadigmJudge] = []

    def add_judge(self, judge: ParadigmJudge) -> ParadigmPanel:
        self._judges.append(judge)
        return self

    def remove_judge(self, name: str) -> ParadigmPanel:
        self._judges = [j for j in self._judges if j.name != name]
        return self

    def get_judge(self, name: str) -> ParadigmJudge | None:
        for j in self._judges:
            if j.name == name:
                return j
        return None

    @property
    def judges(self) -> list[ParadigmJudge]:
        return list(self._judges)

    def evaluate(self, organism: Organism) -> dict[str, Any]:
        if not self._judges:
            return {"scores": {}, "aggregate": 0}
        total_weight = 0.0
        total_score = 0.0
        scores: dict[str, float] = {}
        for judge in self._judges:
            score = judge.evaluate(organism)
            scores[judge.name] = score
            total_score += score * judge.weight
            total_weight += judge.weight
        aggregate = total_score / total_weight if total_weight > 0 else 0
        return {"scores": scores, "aggregate": aggregate}

    def evaluate_population(self, population: list[Organism]) -> list[Organism]:
        for org in population:
            result = self.evaluate(org)
            org.fitness = result["aggregate"]
            org.meta["paradigm_scores"] = result["scores"]
        return population
